deeppavlov/dataset_iterators/multitask_pal_bert_iterator.py

```diff
@@ -0,0 +1,319 @@
+        self.sample_y_instances = None
+
+    def _get_data_size(self, data: Dict[str, List]):
+        """
+        By the way the formula for the sample
+        prob - in the best method - is N^(1-0.8*((epoch - 1)/(NUM_EPOCHS - 1)))
+        so the sample prob is proportional to the number of train examples N
+        """
+        return [len(data[key]) for key in data.keys()]
+
+    def _get_probs(self, data_type):
+        alpha = 1.0 - 0.8 * (self.epochs_done / self.num_train_epochs)
+        data = self._get_data(data_type)
+        probs = [p ** alpha for p in self._get_data_size(data)]
+        tot = sum(probs)
+        probs = [p / tot for p in probs]
+        return probs
+
+    def _extract_data_type(self, data_type):
+        dataset_part = {}
+        for task, iterator in self.task_iterators.items():
+            dataset_part[task] = getattr(iterator, data_type)
+        return dataset_part
+
+    def _get_data(self, data_type):
+        if data_type in self.data.keys():
+            return self.data[data_type]
+        else:
+            raise KeyError(f"Key for {data_type} not found in data")
+
+    @staticmethod
+    def _unite_dataset_parts(*dataset_parts):
+        united = {}
+        for ds_part in dataset_parts:
+            for task, data in ds_part.items():
+                if task not in united:
+                    united[task] = data
+                else:
+                    united[task] = united[task] + data
+        return united
+
+    def _init_sample_tasks_batch(self, batch_size=1):
+        iters = [iter_ for iter_ in self.task_iterators.values()]
+        sample_batch = [i.gen_batches(batch_size).__next__() for i in iters]
+        self.sample_x_instances = []
+        self.sample_y_instances = []
+        for sample_task_batch in sample_batch:
+            self.sample_x_instances.append(sample_task_batch[0])
+            self.sample_y_instances.append(sample_task_batch[1])
+
+    def gen_batches(
+        self, batch_size: int, data_type: str = "train", shuffle: bool = None
+    ) -> Iterator[Tuple[tuple, tuple]]:
+        """Generate batches and expected output to train neural networks. Batches from
+        task iterators are united into one batch. Every element of the largest dataset
+        is used once whereas smaller datasets are repeated until their size is equal to
+        the largest dataset.
+
+        Args:
+            batch_size: number of samples in batch
+            data_type: can be either 'train', 'test', or 'valid'
+            shuffle: whether to shuffle dataset before batching
+
+        Yields:
+            A tuple of a batch of inputs and a batch of expected outputs.
+            Inputs and outputs are tuples. Element of inputs or outputs is a tuple which
+            elements are x values of merged tasks in the ordertasks are present in
+            `tasks` argument of `__init__` method.
+        """
+        max_task_data_len = max([len(iter_.data[data_type])
```

Comment:
I think it's better to just make it a class attribute

Reply:
Made this a class attribute.